flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.runtime.util.EqualityAndHashcodeProvider;
+import org.apache.flink.table.runtime.util.ObjectContainer;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_INTERSECT}. */
+@Internal
+public class ArrayIntersectFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+    private final EqualityAndHashcodeProvider equalityAndHashcodeProvider;
+
+    public ArrayIntersectFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_INTERSECT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        this.equalityAndHashcodeProvider = new EqualityAndHashcodeProvider(context, dataType);
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        equalityAndHashcodeProvider.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData arrayOne, ArrayData arrayTwo) {
+        try {
+            if (arrayOne == null || arrayTwo == null) {
+                return null;
+            }
+
+            List<Object> list = new ArrayList<>();
+            Set<ObjectContainer> set = new HashSet<>();
+            for (int pos = 0; pos < arrayTwo.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayTwo, pos);
+                final ObjectContainer objectContainer = createObjectContainer(element);
+                set.add(objectContainer);
+            }
+            for (int pos = 0; pos < arrayOne.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayOne, pos);
+                final ObjectContainer objectContainer = createObjectContainer(element);
+                if (set.contains(objectContainer)) {
+                    list.add(element);
+                }
+            }
+            return new GenericArrayData(list.toArray());
+        } catch (Throwable t) {
+            throw new FlinkRuntimeException(t);
```

Comment:
From the discussions in [https://github.com/apache/flink/pull/24773](https://github.com/apache/flink/pull/24773) I wonder if we should return null on a runtime exception. For this case null seems to be what is returned if we cannot do an intersect, so it would make send to return a null if there is a runtime error.  I there for suggest catching RuntimeException, logging and returning null.

Reply:
all the builtin function implementation does same @davidradl